compiler/rustc_typeck/src/check/expr.rs

```diff
@@ -1880,21 +1881,103 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+
+        let mut err = type_error_struct!(
             self.tcx().sess,
             span,
             expr_t,
             E0609,
             "no field `{}` on type `{}`",
             field,
             expr_t
-        )
+        );
+
+        // try to add a suggestion in case the field is a nested field of a field of the Adt
+        if let Some((fields, substs)) = self.get_field_candidates(span, &expr_t) {
+            for candidate_field in fields.iter() {
+                if let Some(field_path) =
+                    self.check_for_nested_field(span, field, candidate_field, substs, vec![])
+                {
+                    let base_sp = base_span.with_hi(field.span.lo());
+                    let mut suggestion_str =
+                        self.tcx.sess.source_map().span_to_snippet(base_sp).unwrap();
+                    let field_path_str = field_path
+                        .iter()
+                        .map(|id| id.name.to_ident_string())
+                        .collect::<Vec<String>>()
+                        .join(".");
+                    debug!("field_path_str: {:?}", field_path_str);
+                    suggestion_str.push_str(&field_path_str);
+
+                    err.span_suggestion(
+                        span,
+                        "try to use the whole path to the field",
+                        suggestion_str,
+                        Applicability::MaybeIncorrect,
+                    );
+                }
+            }
+        }
+        err
+    }
+
+    fn get_field_candidates(
+        &self,
+        span: Span,
+        base_t: Ty<'tcx>,
+    ) -> Option<(&Vec<ty::FieldDef>, SubstsRef<'tcx>)> {
+        let mut autoderef = self.autoderef(span, base_t);
+        while let Some((base_t, _)) = autoderef.next() {
+            match base_t.kind() {
+                ty::Adt(base_def, substs) if !base_def.is_enum() => {
+                    let fields = &base_def.non_enum_variant().fields;
+                    return Some((fields, substs));
+                }
+                _ => {}
+            }
+        }
+        None
+    }
+
+    fn check_for_nested_field(
+        &self,
+        span: Span,
+        target_field: Ident,
+        candidate_field: &ty::FieldDef,
+        subst: SubstsRef<'tcx>,
+        field_path: Vec<Ident>,
+    ) -> Option<Vec<Ident>> {
+        let field_ty = candidate_field.ty(self.tcx, subst);
+        if let Some((nested_fields, _)) = self.get_field_candidates(span, &field_ty) {
+            for field in nested_fields.iter() {
+                let mut field_path_clone = field_path.clone();
+                let ident = field.ident.normalize_to_macros_2_0();
+                if ident == target_field {
+                    field_path_clone.push(ident);
+                    return Some(field_path_clone);
+                } else {
+                    field_path_clone.push(ident);
+                    if let Some(path) = self.check_for_nested_field(
```

Comment:
Interesting, I hadn't thought about doing an exhaustive search... We probably want to keep some limits here to avoid accidentally blowing up the compilation time, as this is O(n^2) in the worst case. Maybe up to 4 levels deep and 1000 fields, otherwise early return.

Reply:
It's actually exponential run-time in the worst case. Even limiting the recursion depth and the number of fields to 4 and 1000, resp. still would result in 1000^4 calls in the worst case. I changed this to 3 and 100. 100 might be too low, alternatively we could maybe track and limit the number of recursive calls instead?

Reply:
Yeah, 100 should be fine. Keep in mind that the degenerate case requires all levels of nesting to be structs with 100 fields _each_. A cheeky solution would be to have a _global_ count of fields per level that we pass down (so at any added dot you're always sure you're not going above N total subfields) but that complicates the logic unnecessarily.